src/services/pcn-bridge/src/Bridge.cpp

```diff
@@ -0,0 +1,399 @@
+/**
+ * bridge API generated from bridge.yang
+ *
+ * NOTE: This file is auto generated by polycube-codegen
+ * https://github.com/polycube-network/polycube-codegen
+ */
+
+#include "Bridge.h"
+#include "Bridge_dp_no_stp.h"
+#include "Bridge_dp_stp.h"
+
+Bridge::Bridge(const std::string name, const BridgeJsonObject &conf)
+    : Cube(conf.getBase(),
+           {generate_code(conf.stpEnabledIsSet() ? conf.getStpEnabled()
+                                                 : false)},
+           {}),
+      BridgeBase(name) {
+  logger()->info("Creating Bridge instance");
+
+  if (conf.stpEnabledIsSet())
+    this->stp_enabled_ = conf.getStpEnabled();
+
+  if (conf.macIsSet()) {
+    setMac(conf.getMac());
+  } else {
+    setMac(polycube::service::utils::get_random_mac());
+    logger()->info("[Bridge] Bridge created with random mac: {0}", this->mac_);
+  }
+
+  addPortsList(conf.getPorts());
+  addFdb(conf.getFdb());
+  insertStpList(conf.getStp());
+}
+
+Bridge::~Bridge() {
+  logger()->info("Destroying Bridge instance");
+  this->setStpEnabled(false);
```

Comment:
I added this call not because of the code reloading, but because this would trigger the deletion of PortsStp and Stp instances (and all the threads). In this way, the various components of the bridge are destroyed in order (for each port, firstly PortsStp, then eventually (if the counter is 0) Stp and its thread, and finally Ports instance).
If i don't do this, the `Cube::dismount()` call provokes a fatal error because those components are destroyed not in order (firstly Stp and then Ports instances) and this leads to calls for methods/attributes in objects that are not available anymore.

Maybe i could have adjusted the code in order to prevent this, but i found much simpler to add that line. However, if this is not good, i'll try something else.

Reply:
I got it, I still think that having to reload a program just before deleting the cube is bad, we are wasting time for nothing.
Unfortunately I don't have a possible solution for it yet,

Reply:
I can avoid the reloading of the code.
I added a method called `updatePorts`, that reloads the ports according to a boolean passed as parameter (i just moved the last lines of `setStpEnabled` in this method)

So that part of `setStpEnabled` would become
```
  try {
    reloadCodeWithStp(value);
  } catch (std::exception &e) {
    logger()->error("Failed to reload the code with stp: {0}", value);
  }

  updatePorts(value);
```

And the destructor of Bridge
```
  //setStpEnabled(false); -> remove this line
  if(stp_enabled_)
    updatePorts(false);
```